src/util/permissions.tsx

```diff
@@ -0,0 +1,377 @@
+      }),
+    });
+  }
+
+  resourceOptions.sort(sortOptions);
+
+  return resourceOptions;
+};
+
+export const generateResourceEntitlementOptions = (
+  resourceType: string,
+  permissions?: LxdPermission[],
+) => {
+  if (!permissions || !resourceType) {
+    return {};
+  }
+
+  const resourceEntitlements: Record<
+    string,
+    OptionHTMLAttributes<HTMLOptionElement>[]
+  > = {};
+
+  for (const permission of permissions) {
+    const resource = permission.url;
+    if (!resourceEntitlements[resource]) {
+      resourceEntitlements[resource] = [defaultOption];
+    }
+
+    resourceEntitlements[resource].push({
+      value: permission.entitlement,
+      label: permission.entitlement,
+    });
+  }
+
+  for (const resource in resourceEntitlements) {
+    // split entitlement options into two based on 'can_' prefix
+    const genericEntitlements: OptionHTMLAttributes<HTMLOptionElement>[] = [];
+    const granularEntitlements: OptionHTMLAttributes<HTMLOptionElement>[] = [];
+
+    for (const entitlementOption of resourceEntitlements[resource]) {
+      const entitlementValue = entitlementOption.value as string;
+      if (entitlementValue.includes("can_")) {
+        granularEntitlements.push(entitlementOption);
+        continue;
+      }
+
+      genericEntitlements.push(entitlementOption);
+    }
+
+    genericEntitlements.sort(sortOptions);
+    granularEntitlements.sort(sortOptions);
+
+    // add disabled option as a delimiter between the two sets of entitlements
+    if (genericEntitlements.length > 1 && granularEntitlements.length) {
+      genericEntitlements.splice(1, 0, {
+        disabled: true,
+        label: "Built-in entitlements",
+        value: "",
+      });
+
+      granularEntitlements.unshift({
+        disabled: true,
+        label: "Granular entitlements",
+        value: "",
+      });
+    }
+    resourceEntitlements[resource] =
+      genericEntitlements.concat(granularEntitlements);
+  }
+
+  return resourceEntitlements;
+};
+
+const prefixResourceNameWithProject = (urlSegment: string) => {
+  return (
+    urlSegment.split("?")[1].replace("project=", "") +
+    ": " +
+    urlSegment.split("?")[0]
+  );
+};
+
+// refer to api spec to see how the names can be extracted from resource url
+// https://documentation.ubuntu.com/lxd/en/latest/api/
+export const extractResourceNameFromUrl = (args: {
+  resourceType: string;
+  url: string;
+  imageNamesLookup: Record<string, string>;
+  identityNamesLookup: Record<string, string>;
+}) => {
+  const { resourceType, url, imageNamesLookup, identityNamesLookup } = args;
+  const urlSegments = url.split("/");
+
+  if (resourceType === "server") {
+    return "server";
+  }
+
+  if (resourceType === "identity") {
+    const resourceName = urlSegments[5].split("?")[0];
+    return decodeURIComponent(
+      identityNamesLookup[resourceName] || resourceName,
+    );
+  }
+
+  if (resourceType === "group") {
+    return urlSegments[4].split("?")[0];
+  }
```

Comment:
We should reuse the `filterUsedByType` function and adjust it to this new use case instead of defining a 2nd way of parsing entity urls.

Reply:
We should be more consistent in how we show entities and their details. In Network overview we say "instance-name (project foo)" and leave out the project if it is the default one. Shall we use the same format here as well?


![image](https://github.com/canonical/lxd-ui/assets/1155472/a0c52ea9-4ab9-4bed-b18a-38fb470de31b)

![image](https://github.com/canonical/lxd-ui/assets/1155472/155dc897-4944-448c-a247-f7d6fcfa947a)

Reply:
> Further to the discussion we had, I'm not sure using the same format as the usedBy format makes a lot of sense. In the resource selector, options represent the resource itself and not necessarily the resource that's using it. Take network for example, the instance using the network is not relevant, however we need to show the project for the network because a network can have the same name in different projects.

I think there is a misunderstanding here. What I would like to reuse is parsing the url of a resource. That should be the same logic, independant if the url represents a "used by" entry or a "entitlement resource". In both cases the url defines an entity in lxd.

> Perhaps for the resource options, we can use the bracket format instead of the prefix format?

I'm fine with the bracket format.

Reply:
> I think there is a misunderstanding here. What I would like to reuse is parsing the url of a resource. That should be the same logic, independant if the url represents a "used by" entry or a "entitlement resource". In both cases the url defines an entity in lxd.

Yes I've done exactly that :+1: please have a look at this [diff](https://github.com/canonical/lxd-ui/compare/7011b6260a5b1fe3b1c7d7d351b07dbb24d34fa7..b52ce274253d5a39d5d0ccf3f7720b6149644fe8)